packages/upload-api/src/blob/allocate.js

```diff
@@ -0,0 +1,106 @@
+import * as Server from '@ucanto/server'
+import * as W3sBlob from '@web3-storage/capabilities/web3.storage/blob'
+import * as API from '../types.js'
+import { ensureRateLimitAbove } from '../utils/rate-limits.js'
+
+/**
+ * @param {API.W3ServiceContext} context
+ * @returns {API.ServiceMethod<API.BlobAllocate, API.BlobAllocateSuccess, API.BlobAllocateFailure>}
+ */
+export function blobAllocateProvider(context) {
+  return Server.provide(
+    W3sBlob.allocate,
+    async ({ capability, invocation }) => {
+      const { blob, cause, space } = capability.nb
+      let size = blob.size
+
+      // Rate limiting validation
+      // TODO: we should not produce rate limit error but rather suspend / queue task to be run after enforcing a limit without erroring
+      const rateLimitResult = await ensureRateLimitAbove(
+        context.rateLimitsStorage,
+        [space],
+        0
+      )
+      if (rateLimitResult.error) {
+        return {
+          error: {
+            name: 'RateLimited',
+            message: `${space} is blocked`,
+          },
+        }
+      }
+
+      // Has Storage provider validation
+      const result = await context.provisionsStorage.hasStorageProvider(space)
+      if (result.error) {
+        return result
+      }
+      if (!result.ok) {
+        return {
+          /** @type {API.AllocationError} */
+          error: {
+            name: 'InsufficientStorage',
+            message: `${space} has no storage provider`,
+          },
+        }
+      }
+
+      // Allocate in space, ignoring if already allocated
+      const allocationInsert = await context.allocationsStorage.insert({
+        space,
+        blob,
+        invocation: cause,
+      })
+      if (allocationInsert.error) {
+        // if the insert failed with conflict then this item has already been
+        // added to the space and there is no allocation change.
+        // If record exists but is expired, it can be re-written
+        if (allocationInsert.error.name === 'RecordKeyConflict') {
+          size = 0
+        } else {
+          return {
+            error: allocationInsert.error,
+          }
+        }
+      }
+
+      // Check if blob already exists
+      const hasBlobStore = await context.blobsStorage.has(blob.digest)
+      if (hasBlobStore.error) {
+        return hasBlobStore
+      }
```

Comment:
I don't think we should be checking whether we have a blob during allocation, it should be irrelevant and something we check in the `blob/add` instead.

Reply:
if we have blob we would not want to create a presigned url / in the future not look for a node where to write this, while we want to allocate it to the user. What do you think?